rskj-core/src/main/java/co/rsk/remasc/RemascFederationProvider.java

```diff
@@ -17,44 +17,38 @@
  */
 package co.rsk.remasc;
 
-import co.rsk.config.BridgeConstants;
 import co.rsk.core.RskAddress;
-import co.rsk.peg.BridgeStorageProvider;
+import co.rsk.peg.FederationMember;
 import co.rsk.peg.FederationSupport;
 import org.ethereum.config.blockchain.upgrades.ActivationConfig;
-import org.ethereum.core.Block;
-import org.ethereum.core.Repository;
+import org.ethereum.config.blockchain.upgrades.ConsensusRule;
 import org.ethereum.crypto.ECKey;
-import org.ethereum.vm.PrecompiledContracts;
 
 /**
  * Created by ajlopez on 14/11/2017.
  */
 public class RemascFederationProvider {
+
     private final FederationSupport federationSupport;
+    private final ActivationConfig.ForBlock activations;
 
     public RemascFederationProvider(
-            ActivationConfig activationConfig,
-            BridgeConstants bridgeConstants,
-            Repository repository,
-            Block processingBlock) {
-
-        ActivationConfig.ForBlock activations = activationConfig.forBlock(processingBlock.getNumber());
-        BridgeStorageProvider bridgeStorageProvider = new BridgeStorageProvider(
-                repository,
-                PrecompiledContracts.BRIDGE_ADDR,
-                bridgeConstants,
-                activations
-        );
-        this.federationSupport = new FederationSupport(bridgeConstants, bridgeStorageProvider, processingBlock, activations);
+            ActivationConfig.ForBlock activations,
+            FederationSupport federationSupport) {
+        this.federationSupport = federationSupport;
+        this.activations = activations;
     }
 
     public int getFederationSize() {
         return this.federationSupport.getFederationSize();
     }
 
     public RskAddress getFederatorAddress(int n) {
-        byte[] publicKey = this.federationSupport.getFederatorBtcPublicKey(n);
-        return new RskAddress(ECKey.fromPublicOnly(publicKey).getAddress());
+        if(!activations.isActive(ConsensusRule.RSKIP415)) {
+            byte[] btcPublicKey = this.federationSupport.getFederatorBtcPublicKey(n);
+            return new RskAddress(ECKey.fromPublicOnly(btcPublicKey).getAddress());
+        }
+        byte[] rskPublicKey = this.federationSupport.getFederatorPublicKeyOfType(n, FederationMember.KeyType.RSK);
+        return new RskAddress(ECKey.fromPublicOnly(rskPublicKey).getAddress());

```

Comment:
I have a couple of cleany codey ideas here. What I see is that the whole `return new RskAddress(` statement is repeated. It only changes the public key you provide.

Some ideas
```
        KeyType keyType = KeyType.RSK;
        if(!activations.isActive(ConsensusRule.RSKIP415)) {
            keyType = KeyType.BTC;
        }
        byte[] publicKey = this.federationSupport.getFederatorPublicKeyOfType(n, keyType);
        return new RskAddress(ECKey.fromPublicOnly(publicKey).getAddress());
```
Or 
```
        if(!activations.isActive(ConsensusRule.RSKIP415)) {
            return getRskAddressFromBtcKey();
        }
        return getRskAddressFromRskKey();
```
Maybe the second one is easier tp read but it will result in more duplicated code surely

Reply:
I know is more efficient and it actually makes more sense, but defining the key type as rsk first and then change it is a bit confusing for me.
and maybe the second option is too much... but i think i like it better